tiledb/sm/query/sparse_global_order_reader.cc

```diff
@@ -0,0 +1,918 @@
+          std::vector<uint64_t> coords(dim_num);
+          for (uint32_t d = 0; d < dim_num; ++d) {
+            auto dim = array_schema_->dimension(d);
+            auto rc = ResultCoords(tile, c);
+            coords[d] = dim->map_to_uint64(rc, d, bits, max_bucket_val);
+          }
+
+          // Now we are ready to get the final number.
+          hilbert_values[c] = h.coords_to_hilbert(&coords[0]);
+        }
+        return Status::Ok();
+      });
+
+  RETURN_NOT_OK_ELSE(status, LOG_STATUS(status));
+
+  // Set the values in the comparator.
+  cmp.set_values_for_fragment(frag_idx, hilbert_values);
+
+  return Status::Ok();
+}
+
+Status SparseGlobalOrderReader::compute_tile_result_bitmap(
+    bool subarray_set,
+    ResultTile* tile,
+    std::vector<std::vector<uint8_t>>& tiles_result_bitmap) {
+  // No subarray means we process all cells.
+  if (!subarray_set)
+    return Status::Ok();
+
+  // For easy reference.
+  auto coords_num = tile->cell_num();
+  auto dim_num = array_schema_->dim_num();
+  auto cell_order = array_schema_->cell_order();
+  auto range_coords = subarray_.get_range_coords(0);
+
+  std::vector<uint8_t> result_bitmap(coords_num, 1);
+
+  // Compute result and overwritten bitmap per dimension
+  for (unsigned d = 0; d < dim_num; ++d) {
+    // For col-major cell ordering, iterate the dimensions
+    // in reverse.
+    const unsigned dim_idx =
+        cell_order == Layout::COL_MAJOR ? dim_num - d - 1 : d;
+    const auto& ranges = subarray_.ranges_for_dim(dim_idx);
+    RETURN_NOT_OK(tile->compute_results_sparse(
+        dim_idx, ranges[range_coords[dim_idx]], &result_bitmap, cell_order));
+  }
+
+  tiles_result_bitmap[tile->frag_idx()] = std::move(result_bitmap);
+
+  return Status::Ok();
+}
+
+template <class T>
+Status SparseGlobalOrderReader::compute_result_cell_slabs(
+    uint64_t memory_budget, T cmp) {
+  auto allows_dups = array_schema_->allows_dups();
+
+  // For easy reference.
+  bool subarray_set = subarray_.is_set();
+
+  // A tile min heap, contains one Result coords per fragment.
+  std::priority_queue<ResultCoords, std::vector<ResultCoords>, T> tile_queue(
+      cmp);
+
+  // Tile iterators, per fragments.
+  std::vector<std::list<ResultTile>::iterator> result_tiles_it(
+      result_tiles_.size());
+
+  // Bitmaps for the current tile, per fragment for the cells to process.
+  std::vector<std::vector<uint8_t>> tiles_result_bitmap(result_tiles_.size());
+
+  // For all fragments, get the first tile.
+  for (unsigned int frag_idx = 0; frag_idx < result_tiles_.size(); frag_idx++) {
+    if (result_tiles_[frag_idx].size() > 0) {
+      // Initialize the iterator for this fragment.
+      result_tiles_it[frag_idx] = result_tiles_[frag_idx].begin();
+      auto tile = &*result_tiles_it[frag_idx];
+
+      // Get the cell index we were processing.
+      auto cell_idx = frag_tile_idx_[frag_idx].second;
+
+      // Calculate the bitmap for the cells.
+      auto rc = ResultCoords(tile, cell_idx);
```

Comment:
I am a bit worried about performance here - we may have to move away from `ResultCoords`.

Reply:
Ok I'll have a todo for this.